vowpalwabbit/oaa.cc

```diff
@@ -264,51 +264,70 @@ VW::LEARNER::base_learner* oaa_setup(VW::setup_base_i& stack_builder)
   }
 
   oaa* data_ptr = data.get();
-  VW::LEARNER::learner<oaa, example>* l;
+  uint64_t k_value = data->k;
   auto base = as_singleline(stack_builder.setup_base_learner());
+  void (*learn_ptr)(oaa&, VW::LEARNER::single_learner&, example&);
+  void (*pred_ptr)(oaa&, LEARNER::single_learner&, example&);
+  std::string name_addition;
+  prediction_type_t pred_type;
+  void (*finish_ptr)(vw&, oaa&, example&);
   if (probabilities || scores)
   {
+    pred_type = prediction_type_t::scalars;
     if (probabilities)
     {
       auto loss_function_type = all.loss->getType();
       if (loss_function_type != "logistic")
         *(all.trace_message) << "WARNING: --probabilities should be used only with --loss_function=logistic"
                              << std::endl;
       // the three boolean template parameters are: is_learn, print_all and scores
-      l = &LEARNER::init_multiclass_learner(data, base, learn<false, true, true>, predict<false, true, true>,
-          all.example_parser, data->k, stack_builder.get_setupfn_name(oaa_setup) + "-prob", prediction_type_t::scalars);
-      all.example_parser->lbl_parser.label_type = label_type_t::multiclass;
+      learn_ptr = learn<false, true, true>;
```

Comment:
I am wondering if along with this change, we could start to make parameterization of template functions more explicit.

for example in this case it would look something like
learn_ptr = learn<!Print_All, Scores, Probabilities>

This would really help with functions in gd.

Reply:
I added some directives defining some constants on top as true, so now we can define the function pointers like this:

learn_ptr = learn<!PRINT_ALL, SCORES, PROBABILITIES>;

This should make things more clear